cli-plugins/plugin/plugin.go

```diff
@@ -24,14 +34,57 @@ import (
 // called.
 var PersistentPreRunE func(*cobra.Command, []string) error
 
+// gracefulShutdownTimeout is the time, in seconds, we wait between
+// cancelling a plugins `CommandContext` and forcefully exiting.
+const gracefulShutdownTimeout = 10
+
+// closeOnCLISocketClose connects to the socket specified
+// by the DOCKER_CLI_PLUGIN_SOCKET env var, if present, and attempts
+// to read from it until it receives an EOF, which signals that
+// the CLI is going to exit and the plugin should also exit.
+func closeOnCLISocketClose(errOut io.Writer, cancel func()) {
+	socketAddr, ok := os.LookupEnv(CLIPluginSocketEnvKey)
+	if !ok {
+		// if a plugin compiled against a more recent version of docker/cli
+		// is executed by an older CLI binary, ignore missing environment
+		// variable and behave as usual
+		return
+	}
+	addr, err := net.ResolveUnixAddr("unix", socketAddr)
+	if err != nil {
+		return
+	}
+	cliCloseConn, err := net.DialUnix("unix", nil, addr)
+	if err != nil {
+		return
+	}
+
+	go func() {
+		b := make([]byte, 1)
+		for {
+			_, err := cliCloseConn.Read(b)
+			if errors.Is(err, io.EOF) {
+				cancel()
+				<-time.After(gracefulShutdownTimeout * time.Second)
+				_, _ = fmt.Fprintf(errOut, "plugin did not exit after %d seconds, exiting forcefully\n", gracefulShutdownTimeout)
+				os.Exit(1)
+			}
+		}
+	}()
+}
+
 // RunPlugin executes the specified plugin command
 func RunPlugin(dockerCli *command.DockerCli, plugin *cobra.Command, meta manager.Metadata) error {
 	tcmd := newPluginCommand(dockerCli, plugin, meta)
 
 	var persistentPreRunOnce sync.Once
-	PersistentPreRunE = func(_ *cobra.Command, _ []string) error {
+	PersistentPreRunE = func(cmd *cobra.Command, _ []string) error {
 		var err error
 		persistentPreRunOnce.Do(func() {
+			ctx, cancel := context.WithCancel(cmd.Context())
+			cmd.SetContext(ctx)
```

Comment:
I was curious if `cmd.Context()` would always be set, but it looksthat cobra itself already triggers something to set it, so we should be good (afaics).

Reply:
Hm... although; I see a mention of "can be nil" in https://github.com/docker/cli/pull/4698 (https://github.com/docker/cli/pull/4698/commits/77f2f0fd67ce5fb2789748bcc991aa7496db593e#diff-3dcd6325b70a5f5db393a3ecc08060b69007fcef371fb4dabf3164da95646fb6R80-R82)

So now starting to question myself if there's possible code-paths where it could be nil 🤔

Reply:
I don't think it should happen unless someone calls `cmd.SetContext(nil)`, but we can add a check for that to rule out any possible issues I guess. I'm not sure how we would handle it though – we could either create a context and force it in, but then I imagine if a `cmd.Context` is nil, it might be that way on purpose, or we could skip this whole logic if it is nil.

Reply:
OTOH, we don't really want it to be nil (see #4698), so 🤷‍♀️

Reply:
@neersighted wdyt about this?

Reply:
> context.WithCancel will already panic() if called with nil :) 

Ha! Fair point. Yes my consideration was that if one created a plugin, and no context was set, we should fail hard and loud. But of course that would only make sense if we can confidently check that it's not set.

From a quick check, it looks like Cobra is automatically doing that somewhere in the call tree, which is "good", but of course assuming it _always_ handles this (the comment on the other PR made me a bit concerned if there's code-paths where it would _not_ do so, potentially resulting in a random `panic` 😞.

I don't have a good answer to this one (honestly would've preferred Cobra to fail hard, and require a context to be set at all times if you want to use it, but I guess that complicated things on their side).

Reply:
Looks like I forgot to hit "Comment" yesterday:

I'm personally okay with this as-is; it seems like nullifying the context should be a fairly niche/obviously dubious thing to do, and will result in an appropriate (if potentially opaque) failure.

Reply:
I just missed your comment before pushing 😅. I added some very, very edge case_y_ code to cover this, that shouldn't change anything regardless because if somehow the context was nil before `cmd.Execute()`, Cobra will create a new context for it anyway.